ReCAPTCHA Keys from Options

gamma613 opened the pull request:
Adds a settings page to the plugin, where recaptcha keys can be stored and accessed as opposed to needing to define them each time we add a recaptcha field.

Commit:
Include new settings page

Commit:
add settings page

Commit:
modify to allow settings (site_key, secret_key) from the options table

Commit:
cleanup

Commit:
add settings link to plugins page

irvinlim commented:
Hi, thanks for your PR! I think that in order to maintain backwards compatibility with other ACF reCAPTCHA users, we can't afford to move the settings completely to the settings page.

Perhaps a better option would be to allow users to set default keys, while still being able to override them individually for a single reCAPTCHA field? This also would allow use cases when someone wants to use two different sets of keys.

irvinlim commented on includes/settings-page:
```diff
@@ -0,0 +1,84 @@
+
+    add_settings_section(
+        'acf_recaptcha_settings_section',
+        __('ReCaptcha Keys', 'acf-recaptcha'),
```
Please change the capitalisation of "ReCaptcha" to "reCAPTCHA", to keep it consistent with the rest of the plugin. Same for everywhere.

irvinlim commented on acf-recaptcha-v5.php:
```diff
@@ -65,20 +69,6 @@ function render_field_settings($field) {
-            'name' => 'secret_key',
-            'required' => true,
-        ));
-
```
As mentioned, it would be a good idea to keep the settings for the individual fields.

irvinlim replied:
Also add a note in the instructions that it defaults to the keys specified in the main settings page.

irvinlim commented on acf-recaptcha-v5.php:
```diff
@@ -28,9 +33,8 @@ function __construct() {
+
         $this->defaults = array(
-            'site_key' => '',
-            'secret_key' => '',
```
You can use `get_option('acf_recaptcha')` to populate the defaults.

gamma613 commented:
My only concern with leaving the fields in place is that the same keys will still keep getting saved on each reCAPTCHA field. So if my keys change at some point, I need to update all of my forms again.

Maybe the best solution for the forms would be:

- don't make the site_key or secret_key fields required
- set the ones from the options as a placeholder value (not a default value)
- if the user specifies keys, use those.
- otherwise use the ones from the options table

irvinlim commented:
Yes, that's what I meant - if they leave the individual field settings blank, it will fallback and take the values from the options table.

Commit:
cleanup

- add missing extension to settings page
- standardize the reading of reCAPTCHA
- allow individual settings for reCAPTCHA keys

gamma613 commented:
Requested changes are in place

irvinlim commented on acf-recaptcha-v5.php:
```diff
@@ -69,14 +73,16 @@ function render_field_settings($field) {
             'instructions' => __('Enter your secret key from Google reCAPTCHA.', 'acf-recaptcha'),
             'name' => 'secret_key',
-            'required' => true,
+            'placeholder' => $this->settings['site_key'],
```
I believe this should be `secret_key`.

Anyway maybe we don't want to show the key as the placeholder? Perhaps it could say "Leave blank to use default keys", what do you think?

gamma613 replied:
Good catch. Sorry - that was a case of coding before caffeine. I'm fine with changing it to the placeholder you mentioned.

I had some new ideas relating to these updates. Let me know your thoughts - I might be overthinking

1) if the keys are not present in the options table, make the key fields required. Upon saving, save them to the options table, then null them so they're not saved to the individual field.
2) If ever the user supplies the same keys as we have in the options table, null them so they're not saved to the individual field.

irvinlim replied:
Number 2 may be unnecessary (by [principle of least astonishment](https://www.wikiwand.com/en/Principle_of_least_astonishment)), but we can consider number 1.

Otherwise, it could perhaps just be a notice under `instructions`, informing users that they can add default keys in the settings page as well. Because I actually don't see ACF doing client-side validation if it's omitted - the `required` flag looks like it's just for adding a red asterisk next to the label.

irvinlim commented on includes/settings-page.php:
```diff
@@ -0,0 +1,83 @@
+
+function acf_recaptcha_settings_init()
+{
+    register_setting('acf_recaptcha_settings', 'acf-recaptcha');
```
Likewise, the setting name should be `acf_recaptcha`, I think this is a typo.

gamma613 replied:
I took a look at a number of other plugins I have installed. Most of their settings seem to use - rather than _ as separators for page names. I'm not attached to either way, so if this looks wrong to you I'll change it. This is your plugin after all ;)

irvinlim replied:
Oh, actually it's because the settings can't save, since you're using`get_option(acf_recaptcha)` and the inputs' names are `acf_recaptcha[site_key]`. You'll need to either fix the incorrect option name references, or change this part.

I'm taking reference from https://gist.github.com/annalinneajohansson/5290405, by the way.

irvinlim replied:
I tried it as it was, and nothing was saved to the database until I made this change.

irvinlim reviewed (changes requested):
Just some more nits, and after that I'll just merge and clean up some more things on my end as well.

Commit:
more cleanup

- change setting link to use “_” separator instead of “-“
- make field conditionally required
- conditionally assign placeholders

irvinlim reviewed (approved):
Alright, I'll just take over from here. Thanks for your contribution! 😄

gamma613 commented:
Thanks so much!

irvinlim commented:
No problem! If you wish, do drop the plugin a positive review and give this repository a star if it has helped you :)
